research-software-directory/RSD-as-a-service
# Implements GitLab Scraper

# Implements GitLab Scraper

Closes #216 
Closes #222 
Closes #224

## Frontend
Changes proposed in this pull request:
* Enables user to specify platform of the code repository. This information is required for scrapers to enrich software with number of stats: commit history, programming languages etc.

How to test:
* `docker-compose down --volumes`: remove old version and data
* `docker-compose build`: build all images
* `docker-compose up`: start RSD
* perform data migration: cd data-migration && docker-compose build && docker-compose up
* Login using test account
* Add new software: use plus icon at top right
* Provide Gitlab url, example: https://gitlab.com/dmijatovic/nexter-demo
* After providing repository url the platform will be suggested based on url containing words: `github.`, `gitlab.` or `bitbucket.`
* The user is able to change suggestion by selecting different value in drop-down, below selection "Are you sure?" will appear when selected platform deviates from suggestion (but user selection will be used).
* If you leave edit page and come back later selected platform should be displayed (saved)
* Within few minutes after you provided url to correct Gitlab repo and the platform, the scraper will pull info about programming language and commit history. In the meantime you can validate following...
* View new software page, when Gitlab is defined as platform the Gitlab icon should be displayed.
* When commit history is available the chart will be shown even if get started url is not provided.
* When get started url is provided it will shown the same section including the chart (if commit history is scraper already)

## Screen examples

### Edit software
![image](https://user-images.githubusercontent.com/9204081/163387643-4d3627dc-647b-4a3c-be7a-f5d9ab4103c3.png)

### View software (scraped info)
![image](https://user-images.githubusercontent.com/9204081/163389018-3abba6b8-c870-4e1c-9a13-95844e116c74.png)

## Backend

Changes proposed in this pull request:
* Adds `code_platform` (enum) property to `repository_url` table in the database
* Adds new environment variable `MAX_REQUESTS_GITLAB`
* Adds `GitLabSI` to the scrapers
* Some adjustment to classes used to scrape the data

How to test:

Since there is no support for adding GitLab repositories using the frontend, we need to take a little detour changing the database defaults first:

* in `database/003-create-relations-for-software.sql`, in line 11 (`code_platform`) change default type from `'other'` to `'gitlab'` (alternatively, change `code_platform` manually in the database after adding the repository)
* `docker-compose down --volumes`
* `docker-compose build db scrapers`
* `docker-compose up --force-recreate`
  Use `--force-recreate` if you have docker-compose 2.3, since it may otherwise not recognise that the scrapers container did change
* add a new software with using the repository url https://gitlab.hzdr.de/hifis/overall/surveys/hifis-surveyval
* add a getting-started url (otherwise the graph [will not be displayed](https://github.com/research-software-directory/RSD-as-a-service/pull/217#issuecomment-1098146052))
* wait a few minutes or force the scrapers to run:

  ```
  docker-compose exec scrapers /usr/local/openjdk-17/bin/java -cp /usr/myjava/scrapers.jar nl.esciencecenter.rsd.scraper.MainLicenses
  docker-compose exec scrapers /usr/local/openjdk-17/bin/java -cp /usr/myjava/scrapers.jar nl.esciencecenter.rsd.scraper.MainCommits
  docker-compose exec scrapers /usr/local/openjdk-17/bin/java -cp /usr/myjava/scrapers.jar nl.esciencecenter.rsd.scraper.MainProgrammingLanguages
  ```


PR Checklist:


/edit: update how to test
